peas/sensors.py

```diff
@@ -125,45 +90,103 @@ def capture(self, use_mongo=True, send_message=True):
+        # Note that there is no guarantee that these are the LATEST reports emitted by the sensors,
+        # as the 
         sensor_data = dict()
-
-        # Read from all the readers
         for sensor_name, reader_info in self.serial_readers.items():
             reader = reader_info['reader']
 
-            # Get the values
-            self.logger.debug('Reading next serial value')
-            try:
-                sensor_info = reader.get_reading()
-            except IndexError:
-                continue
-
-            time_stamp = sensor_info[0]
-            sensor_value = sensor_info[1]
+            # Get the values before attempting to re
+            self.logger.debug('ArduinoSerialMonitor.capture reading sensor {}', sensor_name)
             try:
-                self.logger.debug('Got sensor_value from {}'.format(sensor_name))
-                data = yaml.load(sensor_value.replace('nan', 'null'))
+                reading = reader.get_and_parse_reading()
+                if not reading:
+                    self.logger.debug('Unable to get reading from {}', sensor_name)
+                    continue
+                self.logger.debug('Got sensor_value from {}', sensor_name)
+                time_stamp, data = reading
                 data['date'] = time_stamp
-
                 sensor_data[sensor_name] = data
-
                 if send_message:
                     self.send_message({'data': data}, channel='environment')
-            except yaml.parser.ParserError:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
-            except ValueError:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
-            except TypeError:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
             except Exception as e:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
+                self.logger.warning('Exception while reading from sensor {}: {}', sensor_name, e)
 
-            if use_mongo and len(sensor_data) > 0:
-                if self.db is None:
-                    self.db = PanMongo()
-                    self.logger.info('Connected to PanMongo')
-                self.db.insert_current('environment', sensor_data)
-        else:
-            self.logger.debug('No sensor data received')
+        if use_mongo and len(sensor_data) > 0:
+            if self.db is None:
+                self.db = PanMongo()
+                self.logger.info('Connected to PanMongo')
+            self.db.insert_current('environment', sensor_data)
 
         return sensor_data
+
+
+def auto_detect_arduino_devices(comports=None, logger=None):
+    if comports is None:
+        comports = find_arduino_devices()
+    if not logger:
+        logger = get_root_logger()
+    result = []
+    for port in comports:
+        v = auto_detect_port(port, logger)
+        if v:
+            result.append(v)
+    return result
+
+
+def find_arduino_devices():
```

Comment:
This probably makes the most sense here but just wanted to point out that we have a [`list_connected_cameras`](
https://github.com/panoptes/POCS/blob/develop/pocs/utils/__init__.py#L74) function in the utils. We should aim for some consistency of putting these kinds of functions either in the file like you have here or create a whole device discovery utility.

Reply:
Good idea. I'll add a comment or issue to that effect.

Reply:
#339 entered.